library/init/meta/lean/parser.lean

```diff
@@ -85,5 +107,19 @@ meta constant of_tactic : tactic α → parser α
 meta instance : has_coe (tactic α) (parser α) :=
 ⟨of_tactic⟩
 
+namespace reflectable
+
+meta instance cast (p : lean.parser (reflected_value α)) : reflectable (val p) :=
+{full := p}
+
+meta instance has_reflect [r : has_reflect α] (p : lean.parser α) : reflectable p :=
+{full := do rp ← p, return (reflected_value.reflect rp)}
+
+meta instance optional {α : Type} [reflected α] {p : parser (reflected_value α)}
+    : reflectable (optional (val p)) :=
+{full := reflected_value.subst some <$> p <|> return (reflected_value.mk none)}
```

Comment:
I think you should have an instance of `reflectable (val p)` that uses `p`... That way you can have the `optional` instance say `[reflectable p] : reflectable (optional p)`, which is what I expected to see. Or is there another reason you wrote it that way?

Reply:
That is instance is the content of `reflectable.cast`, and sure that sounds better.